azure-kit-starter/src/main/java/com/vaadin/azure/starter/ui/ClusterSupport.java

```diff
@@ -0,0 +1,174 @@
+        serviceInitEvent.addRequestHandler(createRequestHandler());
+    }
+
+    private RequestHandler createRequestHandler() {
+        return (vaadinSession, vaadinRequest, vaadinResponse) -> {
+            // Set the thread local instance
+            CurrentInstance.set(ClusterSupport.class, this);
+
+            vaadinSession.access(handleVersions(vaadinSession, vaadinResponse));
+
+            // If the current and the new versions do not match notify the user
+            return false;
+        };
+    }
+
+    private Command handleVersions(VaadinSession vaadinSession,
+            VaadinResponse vaadinResponse) {
+        return () -> {
+            // Always set the version cookie
+            Cookie versionCookie = getCookieByName(CURRENT_VERSION_COOKIE);
+            if (versionCookie == null
+                    || !versionCookie.getValue().equals(appVersion)) {
+                versionCookie = new Cookie(CURRENT_VERSION_COOKIE, appVersion);
+                versionCookie.setHttpOnly(true);
+                vaadinResponse.addCookie(versionCookie);
+            }
+
+            // Always check for the new version cookie
+            Cookie currentCookie = getCookieByName(UPDATE_VERSION_COOKIE);
+            if (currentCookie != null && !currentCookie.getValue().isEmpty()
+                    && !versionCookie.getValue()
+                            .equals(currentCookie.getValue())) {
+                vaadinSession.getUIs().forEach(ui -> {
+                    if (ui.getChildren().anyMatch(
+                            child -> (child instanceof VersionNotificator))) {
+                        return;
+                    }
+                    VersionNotificator notificator = new VersionNotificator(
+                            appVersion, currentCookie.getValue());
+                    notificator.addSwitchVersionEventListener(
+                            createSwitchVersionEventListener());
+                    // Show notificator
+                    ui.add(notificator);
+                });
+            }
+        };
+    }
+
+    private ComponentEventListener<VersionNotificator.SwitchVersionEvent> createSwitchVersionEventListener() {
+        return e -> {
+            // Do nothing if switch version listener prevents switching
+            if (switchVersionListener != null && !switchVersionListener
+                    .clusterSwitch(VaadinRequest.getCurrent(),
+                            VaadinResponse.getCurrent())) {
+                return;
+            }
+
+            // When the user clicks on the notificator remove the cluster
+            // and session cookies
+            removeCookie(STICKY_CLUSTER_COOKIE);
+            removeCookie(CURRENT_VERSION_COOKIE);
+            removeCookie(UPDATE_VERSION_COOKIE);
+
+            // Invalidate the session, vaadin does not synchronizes it
+            // between clusters
+            VaadinRequest.getCurrent().getWrappedSession().invalidate();
+        };
+    }
```

Comment:
Same here: make the method be the listener itself, so we save an indentation level.
```suggestion
                    notificator.addSwitchVersionEventListener(
                            this::handleSwitchVersionEvent);
                    // Show notificator
                    ui.add(notificator);
                });
            }
        };
    }

    private void handleSwitchVersionEvent(
            VersionNotificator.SwitchVersionEvent event) {
        // Do nothing if switch version listener prevents switching
        if (switchVersionListener != null && !switchVersionListener
                .clusterSwitch(VaadinRequest.getCurrent(),
                        VaadinResponse.getCurrent())) {
            return;
        }

        // When the user clicks on the notificator remove the cluster
        // and session cookies
        removeCookie(STICKY_CLUSTER_COOKIE);
        removeCookie(CURRENT_VERSION_COOKIE);
        removeCookie(UPDATE_VERSION_COOKIE);

        // Invalidate the session, vaadin does not synchronizes it
        // between clusters
        VaadinRequest.getCurrent().getWrappedSession().invalidate();
    }
```

Reply:
Refactored to use method reference.